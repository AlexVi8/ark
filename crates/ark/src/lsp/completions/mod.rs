//
// mod.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

mod completion_item;
mod document;
mod provide;
mod resolve;
mod session;
mod types;
mod workspace;

pub use provide::provide_completions;
pub use resolve::resolve_completion;
