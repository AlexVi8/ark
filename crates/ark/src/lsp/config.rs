use serde::Deserialize;
use serde::Serialize;
use serde_json::Value;

use crate::lsp::diagnostics::DiagnosticsConfig;

pub struct Setting<T> {
    pub key: &'static str,
    pub set: fn(&mut T, Value),
}

/// List of LSP settings for which clients can send `didChangeConfiguration`
/// notifications. We register our interest in watching over these settings in
/// our `initialized` handler. The `set` methods convert from a json `Value` to
/// the expected type, using a default value if the conversion fails.
///
/// This array is for global settings. If the setting should only affect a given
/// document URI, add it to `DOCUMENT_SETTINGS` instead.
pub static GLOBAL_SETTINGS: &[Setting<LspConfig>] = &[
    Setting {
        key: "positron.r.diagnostics.enable",
        set: |cfg, v| {
            cfg.diagnostics.enable = v
                .as_bool()
                .unwrap_or_else(|| DiagnosticsConfig::default().enable)
        },
    },
    Setting {
        key: "positron.r.symbols.includeAssignmentsInBlocks",
        set: |cfg, v| {
            cfg.symbols.include_assignments_in_blocks = v
                .as_bool()
                .unwrap_or_else(|| SymbolsConfig::default().include_assignments_in_blocks)
        },
    },
    Setting {
        key: "positron.r.workspaceSymbols.includeCommentSections",
        set: |cfg, v| {
            cfg.workspace_symbols.include_comment_sections = v
                .as_bool()
                .unwrap_or_else(|| WorkspaceSymbolsConfig::default().include_comment_sections)
        },
    },
];

/// These document settings are updated on a URI basis. Each document has its
/// own value of the setting.
pub static DOCUMENT_SETTINGS: &[Setting<DocumentConfig>] = &[
    Setting {
        key: "editor.insertSpaces",
        set: |cfg, v| {
            let default_style = IndentationConfig::default().indent_style;
            cfg.indent.indent_style = if v
                .as_bool()
                .unwrap_or_else(|| default_style == IndentStyle::Space)
            {
                IndentStyle::Space
            } else {
                IndentStyle::Tab
            }
        },
    },
    Setting {
        key: "editor.indentSize",
        set: |cfg, v| {
            cfg.indent.indent_size = v
                .as_u64()
                .map(|n| n as usize)
                .unwrap_or_else(|| IndentationConfig::default().indent_size)
        },
    },
    Setting {
        key: "editor.tabSize",
        set: |cfg, v| {
            cfg.indent.tab_width = v
                .as_u64()
                .map(|n| n as usize)
                .unwrap_or_else(|| IndentationConfig::default().tab_width)
        },
    },
];

/// Configuration of the LSP
#[derive(Clone, Default, Debug)]
pub(crate) struct LspConfig {
    pub(crate) diagnostics: DiagnosticsConfig,
    pub(crate) symbols: SymbolsConfig,
    pub(crate) workspace_symbols: WorkspaceSymbolsConfig,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct SymbolsConfig {
    /// Whether to emit assignments in `{` bloks as document symbols.
    pub include_assignments_in_blocks: bool,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct WorkspaceSymbolsConfig {
    /// Whether to include sections like `# My section ---` in workspace symbols.
    pub include_comment_sections: bool,
}

/// Configuration of a document.
///
/// The naming follows <https://editorconfig.org/> where possible.
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct DocumentConfig {
    pub indent: IndentationConfig,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct IndentationConfig {
    /// Whether to insert spaces of tabs for one level of indentation.
    pub indent_style: IndentStyle,

    /// The number of spaces for one level of indentation.
    pub indent_size: usize,

    /// The width of a tab. There may be projects with an `indent_size` of 4 and
    /// a `tab_width` of 8 (e.g. GNU R).
    pub tab_width: usize,
}

#[derive(PartialEq, Serialize, Deserialize, Clone, Debug)]
pub enum IndentStyle {
    Tab,
    Space,
}

impl Default for SymbolsConfig {
    fn default() -> Self {
        Self {
            include_assignments_in_blocks: false,
        }
    }
}

impl Default for WorkspaceSymbolsConfig {
    fn default() -> Self {
        Self {
            include_comment_sections: false,
        }
    }
}

impl Default for IndentationConfig {
    fn default() -> Self {
        Self {
            indent_style: IndentStyle::Space,
            indent_size: 2,
            tab_width: 2,
        }
    }
}

pub(crate) fn indent_style_from_lsp(insert_spaces: bool) -> IndentStyle {
    if insert_spaces {
        IndentStyle::Space
    } else {
        IndentStyle::Tab
    }
}
