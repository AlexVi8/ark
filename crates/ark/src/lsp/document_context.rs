//
// document_context.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use tree_sitter::Node;
use tree_sitter::Point;

use crate::lsp::documents::Document;
use crate::lsp::traits::node::NodeExt;

#[derive(Debug)]
pub struct DocumentContext<'a> {
    pub document: &'a Document,
    pub node: Node<'a>,
    pub source: String,
    pub point: Point,
}

impl<'a> DocumentContext<'a> {
    pub fn new(document: &'a Document, point: Point) -> Self {
        // get reference to AST
        let ast = &document.ast;

        // find node at point
        let node = ast.root_node().find_closest_node_to_point(point).unwrap();

        // convert document contents to a string once, to be reused elsewhere
        let source = document.contents.to_string();

        // build document context
        DocumentContext {
            document,
            node,
            source,
            point,
        }
    }
}
