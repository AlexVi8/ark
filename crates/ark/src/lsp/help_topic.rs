//
// help_topic.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use serde::Deserialize;
use serde::Serialize;
use tower_lsp::lsp_types::Position;
use tower_lsp::lsp_types::VersionedTextDocumentIdentifier;

use crate::backend_trace;
use crate::lsp::backend::Backend;
use crate::lsp::traits::cursor::TreeCursorExt;
use crate::lsp::traits::position::PositionExt;

pub static POSITRON_HELP_TOPIC_REQUEST: &'static str = "positron/textDocument/helpTopic";

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HelpTopicParams {
    /// The document to provide a help topic for.
    pub text_document: VersionedTextDocumentIdentifier,
    /// The location of the cursor.
    pub position: Position,
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HelpTopicResponse {
    /// The help topic appropriate for the cursor position.
    pub topic: String,
}

impl Backend {
    pub async fn help_topic(
        &self,
        params: HelpTopicParams,
    ) -> tower_lsp::jsonrpc::Result<Option<HelpTopicResponse>> {
        backend_trace!(self, "help_topic({:?})", params);

        let uri = &params.text_document.uri;
        let Some(document) = self.documents.get_mut(uri) else {
            backend_trace!(self, "help_topic(): No document associated with URI {uri}");
            return Ok(None);
        };

        let root = document.ast.root_node();
        let position = params.position;
        let point = position.as_point();

        let mut cursor = root.walk();
        let node = cursor.find_leaf(point);

        // Check to see whether it's a function node (currently that's what we
        // support)
        if node.kind() != "function" {
            backend_trace!(self, "help_topic(): No function node at position {point}");
            return Ok(None);
        }

        // Get the text of the node
        let source = document.contents.to_string();
        let text = node.utf8_text(source.as_bytes()).unwrap();

        let response = HelpTopicResponse {
            topic: String::from(text),
        };

        Ok(Some(response))
    }
}
