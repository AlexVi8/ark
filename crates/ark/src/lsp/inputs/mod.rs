//
// mod.rs
//
// Copyright (C) 2025 by Posit Software, PBC
//
//

pub mod documentation_rd_file;
pub mod library;
pub mod package;
pub mod package_description;
pub mod package_namespace;
pub mod source_root;
