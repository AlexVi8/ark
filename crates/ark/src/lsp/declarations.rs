use tree_sitter::Node;

use crate::lsp;
use crate::treesitter::args_find_call_args;
use crate::treesitter::node_arg_value;
use crate::treesitter::node_is_call;
use crate::treesitter::node_text;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;
use crate::treesitter::UnaryOperatorType;

pub(crate) struct TopLevelDeclare {
    pub(crate) diagnostics: bool,
}

impl Default for TopLevelDeclare {
    fn default() -> Self {
        Self { diagnostics: true }
    }
}

pub(crate) fn top_level_declare(
    ast: &tree_sitter::Tree,
    contents: &ropey::Rope,
) -> TopLevelDeclare {
    let mut decls = TopLevelDeclare::default();

    let Some(declare_args) = top_level_declare_args(ast, contents) else {
        return decls;
    };
    let Some(args) = declare_ark_args(declare_args, contents) else {
        return decls;
    };

    let mut cursor = args.walk();
    let mut iter = args.children(&mut cursor);

    let Some(diags) = iter.find_map(|n| node_arg_value(&n, "diagnostics", contents)) else {
        return decls;
    };
    let Some(diags_text) = node_text(&diags, contents) else {
        return decls;
    };

    if diags_text == "FALSE" {
        decls.diagnostics = false;
    } else if diags_text != "TRUE" {
        lsp::log_warn!("Invalid `diagnostics = ` declaration");
    }

    decls
}

fn top_level_declare_args<'tree>(
    ast: &'tree tree_sitter::Tree,
    contents: &ropey::Rope,
) -> Option<Node<'tree>> {
    let root = ast.root_node();
    let mut cursor = root.walk();
    let iter = root.children(&mut cursor);

    // The declarations are allowed to appear after top comments
    let mut iter = iter.skip_while(|n| n.is_comment());
    let Some(mut first) = iter.next() else {
        return None;
    };

    // For backward compatibility with R < 4.4.0, declarations may be wrapped in
    // a tilde call
    if first.node_type() == NodeType::UnaryOperator(UnaryOperatorType::Tilde) {
        first = first.child_by_field_name("rhs")?;
    }

    if !node_is_call(&first, "declare", contents) {
        return None;
    }

    first.child_by_field_name("arguments")
}

fn declare_ark_args<'tree>(
    declare_args: Node<'tree>,
    contents: &ropey::Rope,
) -> Option<Node<'tree>> {
    args_find_call_args(declare_args, "ark", contents)
}

#[cfg(test)]
mod test {
    use harp::assert_match;

    use crate::lsp::declarations::declare_ark_args;
    use crate::lsp::declarations::top_level_declare;
    use crate::lsp::declarations::top_level_declare_args;
    use crate::lsp::documents::Document;

    #[test]
    fn test_declare_args() {
        let doc = Document::new("", None);
        assert_match!(top_level_declare_args(&doc.ast, &doc.contents), None);

        let doc = Document::new("declare()", None);
        assert_match!(top_level_declare_args(&doc.ast, &doc.contents), Some(_));

        let doc = Document::new("~declare()", None);
        assert_match!(top_level_declare_args(&doc.ast, &doc.contents), Some(_));

        let doc = Document::new("# foo\n#bar\n\ndeclare()", None);
        assert_match!(top_level_declare_args(&doc.ast, &doc.contents), Some(_));

        let doc = Document::new("# foo\nbar\n\ndeclare()", None);
        assert_match!(top_level_declare_args(&doc.ast, &doc.contents), None);
    }

    #[test]
    fn test_declare_ark_args() {
        let doc = Document::new("declare()", None);
        let decls = top_level_declare_args(&doc.ast, &doc.contents).unwrap();
        assert_match!(declare_ark_args(decls, &doc.contents), None);

        let doc = Document::new("declare(ark())", None);
        let decls = top_level_declare_args(&doc.ast, &doc.contents).unwrap();
        assert_match!(declare_ark_args(decls, &doc.contents), Some(_));

        let doc = Document::new("declare(foo, ark())", None);
        let decls = top_level_declare_args(&doc.ast, &doc.contents).unwrap();
        assert_match!(declare_ark_args(decls, &doc.contents), Some(_));
    }

    #[test]
    fn test_declare_diagnostics() {
        let doc = Document::new("", None);
        let decls = top_level_declare(&doc.ast, &doc.contents);
        assert_eq!(decls.diagnostics, true);

        let doc = Document::new("declare(ark(diagnostics = TRUE))", None);
        let decls = top_level_declare(&doc.ast, &doc.contents);
        assert_eq!(decls.diagnostics, true);

        let doc = Document::new("declare(ark(diagnostics = NULL))", None);
        let decls = top_level_declare(&doc.ast, &doc.contents);
        assert_eq!(decls.diagnostics, true);

        let doc = Document::new("declare(ark(diagnostics = invalid()))", None);
        let decls = top_level_declare(&doc.ast, &doc.contents);
        assert_eq!(decls.diagnostics, true);

        let doc = Document::new("~declare()", None);
        let decls = top_level_declare(&doc.ast, &doc.contents);
        assert_eq!(decls.diagnostics, true);

        let doc = Document::new("declare(ark(diagnostics = FALSE))", None);
        let decls = top_level_declare(&doc.ast, &doc.contents);
        assert_eq!(decls.diagnostics, false);
    }
}
