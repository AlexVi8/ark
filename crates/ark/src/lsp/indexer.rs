//
// indexer.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use std::collections::HashMap;
use std::path::Path;
use std::result::Result::Ok;
use std::sync::Arc;
use std::sync::LazyLock;
use std::sync::Mutex;

use anyhow::anyhow;
use regex::Regex;
use ropey::Rope;
use stdext::unwrap;
use stdext::unwrap::IntoResult;
use tower_lsp::lsp_types::Range;
use tree_sitter::Node;
use walkdir::DirEntry;
use walkdir::WalkDir;

use crate::lsp;
use crate::lsp::documents::Document;
use crate::lsp::encoding::convert_point_to_position;
use crate::lsp::traits::rope::RopeExt;
use crate::treesitter::BinaryOperatorType;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;

#[derive(Clone, Debug)]
pub enum IndexEntryData {
    Variable {
        name: String,
    },
    Function {
        name: String,
        arguments: Vec<String>,
    },
    // Like Function but not used for completions yet
    Method {
        name: String,
    },
    Section {
        level: usize,
        title: String,
    },
}

#[derive(Clone, Debug)]
pub struct IndexEntry {
    pub key: String,
    pub range: Range,
    pub data: IndexEntryData,
}

type DocumentPath = String;
type DocumentSymbol = String;
type DocumentSymbolIndex = HashMap<DocumentSymbol, IndexEntry>;
type WorkspaceIndex = Arc<Mutex<HashMap<DocumentPath, DocumentSymbolIndex>>>;

static WORKSPACE_INDEX: LazyLock<WorkspaceIndex> = LazyLock::new(|| Default::default());
pub static RE_COMMENT_SECTION: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\s*(#+)\s*(.*?)\s*[#=-]{4,}\s*$").unwrap());

#[tracing::instrument(level = "info", skip_all)]
pub fn start(folders: Vec<String>) {
    let now = std::time::Instant::now();
    lsp::log_info!("Initial indexing started");

    for folder in folders {
        let walker = WalkDir::new(folder);
        for entry in walker.into_iter().filter_entry(|e| filter_entry(e)) {
            if let Ok(entry) = entry {
                if entry.file_type().is_file() {
                    if let Err(err) = index_file(entry.path()) {
                        lsp::log_error!("Can't index file {:?}: {err:?}", entry.path());
                    }
                }
            }
        }
    }

    lsp::log_info!(
        "Initial indexing finished after {}ms",
        now.elapsed().as_millis()
    );
}

pub fn find(symbol: &str) -> Option<(String, IndexEntry)> {
    let index = WORKSPACE_INDEX.lock().unwrap();

    for (path, index) in index.iter() {
        if let Some(entry) = index.get(symbol) {
            return Some((path.clone(), entry.clone()));
        }
    }

    None
}

pub fn map(mut callback: impl FnMut(&Path, &String, &IndexEntry)) {
    let index = WORKSPACE_INDEX.lock().unwrap();

    for (path, index) in index.iter() {
        for (symbol, entry) in index.iter() {
            let path = Path::new(path);
            callback(path, symbol, entry);
        }
    }
}

#[tracing::instrument(level = "trace", skip_all, fields(path = ?path))]
pub fn update(document: &Document, path: &Path) -> anyhow::Result<()> {
    clear(path)?;
    index_document(document, path);
    Ok(())
}

fn insert(path: &Path, entry: IndexEntry) -> anyhow::Result<()> {
    let mut index = WORKSPACE_INDEX.lock().unwrap();
    let path = str_from_path(path)?;

    let index = index.entry(path.to_string()).or_default();

    // Retain the first occurrence in the index. In the future we'll track every occurrences and
    // their scopes but for now we only track the first definition of an object (in a way, its
    // declaration).
    if !index.contains_key(&entry.key) {
        index.insert(entry.key.clone(), entry);
    }

    Ok(())
}

fn clear(path: &Path) -> anyhow::Result<()> {
    let mut index = WORKSPACE_INDEX.lock().unwrap();
    let path = str_from_path(path)?;

    // Only clears if the `path` was an existing key
    index.entry(path.into()).and_modify(|index| {
        index.clear();
    });

    Ok(())
}

fn str_from_path(path: &Path) -> anyhow::Result<&str> {
    path.to_str().ok_or(anyhow!(
        "Couldn't convert path {} to string",
        path.display()
    ))
}

// TODO: Should we consult the project .gitignore for ignored files?
// TODO: What about front-end ignores?
// TODO: What about other kinds of ignores (e.g. revdepcheck)?
pub fn filter_entry(entry: &DirEntry) -> bool {
    let name = entry.file_name();

    // skip common ignores
    for ignore in [".git", ".Rproj.user", "node_modules", "revdep"] {
        if name == ignore {
            return false;
        }
    }

    // skip project 'renv' folder
    if name == "renv" {
        let companion = entry.path().join("activate.R");
        if companion.exists() {
            return false;
        }
    }

    true
}

fn index_file(path: &Path) -> anyhow::Result<()> {
    // only index R files
    let ext = path.extension().unwrap_or_default();
    if ext != "r" && ext != "R" {
        return Ok(());
    }

    // TODO: Handle document encodings here.
    // TODO: Check if there's an up-to-date buffer to be used.
    let contents = std::fs::read(path)?;
    let contents = String::from_utf8(contents)?;
    let document = Document::new(contents.as_str(), None);

    index_document(&document, path);

    Ok(())
}

fn index_document(document: &Document, path: &Path) {
    let ast = &document.ast;
    let contents = &document.contents;

    let root = ast.root_node();
    let mut cursor = root.walk();
    let mut entries = Vec::new();

    for node in root.children(&mut cursor) {
        if let Err(err) = index_node(path, contents, &node, &mut entries) {
            lsp::log_error!("Can't index document: {err:?}");
        }
    }

    for entry in entries {
        if let Err(err) = insert(path, entry) {
            lsp::log_error!("Can't insert index entry: {err:?}");
        }
    }
}

fn index_node(
    path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    index_assignment(path, contents, node, entries)?;
    index_comment(path, contents, node, entries)?;
    Ok(())
}

fn index_assignment(
    path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    if !matches!(
        node.node_type(),
        NodeType::BinaryOperator(BinaryOperatorType::LeftAssignment) |
            NodeType::BinaryOperator(BinaryOperatorType::EqualsAssignment)
    ) {
        return Ok(());
    }

    let lhs = match node.child_by_field_name("lhs") {
        Some(lhs) => lhs,
        None => return Ok(()),
    };

    let Some(rhs) = node.child_by_field_name("rhs") else {
        return Ok(());
    };

    if crate::treesitter::node_is_call(&rhs, "R6Class", contents) ||
        crate::treesitter::node_is_namespaced_call(&rhs, "R6", "R6Class", contents)
    {
        index_r6_class(path, contents, &rhs, entries)?;
    }

    let lhs_text = contents.node_slice(&lhs)?.to_string();

    // The method matching is super hacky but let's wait until the typed API to
    // do better
    if !lhs_text.starts_with("method(") && !lhs.is_identifier_or_string() {
        return Ok(());
    }

    let Some(rhs) = node.child_by_field_name("rhs") else {
        return Ok(());
    };

    if rhs.is_function_definition() {
        // If RHS is a function definition, emit a function symbol
        let mut arguments = Vec::new();
        if let Some(parameters) = rhs.child_by_field_name("parameters") {
            let mut cursor = parameters.walk();
            for child in parameters.children(&mut cursor) {
                let name = unwrap!(child.child_by_field_name("name"), None => continue);
                if name.is_identifier() {
                    let name = contents.node_slice(&name)?.to_string();
                    arguments.push(name);
                }
            }
        }

        let start = convert_point_to_position(contents, lhs.start_position());
        let end = convert_point_to_position(contents, lhs.end_position());

        entries.push(IndexEntry {
            key: lhs_text.clone(),
            range: Range { start, end },
            data: IndexEntryData::Function {
                name: lhs_text,
                arguments,
            },
        });
    } else {
        // Otherwise, emit variable
        let start = convert_point_to_position(contents, lhs.start_position());
        let end = convert_point_to_position(contents, lhs.end_position());
        entries.push(IndexEntry {
            key: lhs_text.clone(),
            range: Range { start, end },
            data: IndexEntryData::Variable { name: lhs_text },
        });
    }

    Ok(())
}

fn index_r6_class(
    _path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    let Some(args_node) = node.child_by_field_name("arguments") else {
        return Ok(());
    };

    let mut cursor = args_node.walk();
    for arg in args_node.children(&mut cursor) {
        // Only consider `public = ` and `private = ` arguments
        let Some(arg_name) = arg.child_by_field_name("name") else {
            continue;
        };
        if !arg_name.is_identifier_or_string() {
            continue;
        }
        let arg_name_str = contents.node_slice(&arg_name)?;
        if arg_name_str != "public" && arg_name_str != "private" {
            continue;
        }

        let Some(list_node) = arg.child_by_field_name("value") else {
            continue;
        };
        if !list_node.is_call() {
            continue;
        }

        let Some(list_args) = list_node.child_by_field_name("arguments") else {
            return Ok(());
        };

        let mut cursor = list_args.walk();
        for arg in list_args.children(&mut cursor) {
            if !arg.is_argument() {
                continue;
            }

            let (Some(mtd_name), Some(mtd_value)) = (
                arg.child_by_field_name("name"),
                arg.child_by_field_name("value"),
            ) else {
                continue;
            };
            if !mtd_name.is_identifier_or_string() || !mtd_value.is_function_definition() {
                continue;
            }

            let name = contents.node_slice(&mtd_name)?.to_string();
            let start = convert_point_to_position(contents, mtd_name.start_position());
            let end = convert_point_to_position(contents, mtd_name.end_position());

            entries.push(IndexEntry {
                key: name.clone(),
                range: Range { start, end },
                data: IndexEntryData::Method { name },
            });
        }
    }

    Ok(())
}

fn index_comment(
    _path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    // check for comment
    if !node.is_comment() {
        return Ok(());
    }

    // see if it looks like a section
    let comment = contents.node_slice(node)?.to_string();
    let matches = match RE_COMMENT_SECTION.captures(comment.as_str()) {
        Some(m) => m,
        None => return Ok(()),
    };

    let level = matches.get(1).into_result()?;
    let title = matches.get(2).into_result()?;

    let level = level.as_str().len();
    let title = title.as_str().to_string();

    // skip things that look like knitr output
    if title.starts_with("----") {
        return Ok(());
    }

    let start = convert_point_to_position(contents, node.start_position());
    let end = convert_point_to_position(contents, node.end_position());

    entries.push(IndexEntry {
        key: title.clone(),
        range: Range::new(start, end),
        data: IndexEntryData::Section { level, title },
    });

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::assert_debug_snapshot;

    use super::*;
    use crate::lsp::documents::Document;

    macro_rules! test_index {
        ($code:expr) => {
            let doc = Document::new($code, None);
            let path = PathBuf::from("/path/to/file.R");
            let root = doc.ast.root_node();
            let mut cursor = root.walk();

            let mut entries = vec![];
            for node in root.children(&mut cursor) {
                let _ = index_node(&path, &doc.contents, &node, &mut entries);
            }
            assert_debug_snapshot!(entries);
        };
    }

    // Note that unlike document symbols whose ranges cover the whole entity
    // they represent, the range of workspace symbols only cover the identifers

    #[test]
    fn test_index_function() {
        test_index!(
            r#"
my_function <- function(a, b = 1) {
  a + b

  # These are not indexed as workspace symbol
  inner <- function() {
    2
  }
  inner_var <- 3
}

my_variable <- 1
"#
        );
    }

    #[test]
    fn test_index_variable() {
        test_index!(
            r#"
x <- 10
y = "hello"
"#
        );
    }

    #[test]
    fn test_index_s7_methods() {
        test_index!(
            r#"
Class <- new_class("Class")
generic <- new_generic("generic", "arg",
  function(arg) {
    S7_dispatch()
  }
)
method(generic, Class) <- function(arg) {
  NULL
}
"#
        );
    }

    #[test]
    fn test_index_comment_section() {
        test_index!(
            r#"
# Section 1 ----
x <- 10

## Subsection ======
y <- 20

x <- function() {
    # This inner section is not indexed ----
}

"#
        );
    }

    #[test]
    fn test_index_r6class() {
        test_index!(
            r#"
class <- R6Class(
    public = list(
        initialize = function() {
            1
        },
        public_method = function() {
            2
        }
    ),
    private = list(
        private_method = function() {
            1
        }
    ),
    other = list(
        other_method = function() {
            1
        }
    )
)
"#
        );
    }

    #[test]
    fn test_index_r6class_namespaced() {
        test_index!(
            r#"
class <- R6::R6Class(
    public = list(
        initialize = function() {
            1
        },
    )
)
"#
        );
    }
}
