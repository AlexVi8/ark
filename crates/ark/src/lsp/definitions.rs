//
// definitions.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use anyhow::Result;
use tower_lsp::lsp_types::GotoDefinitionParams;
use tower_lsp::lsp_types::GotoDefinitionResponse;
use tower_lsp::lsp_types::LocationLink;
use tower_lsp::lsp_types::Range;
use tower_lsp::lsp_types::Url;

use crate::lsp::documents::Document;
use crate::lsp::encoding::convert_point_to_position;
use crate::lsp::encoding::convert_position_to_point;
use crate::lsp::indexer;
use crate::lsp::traits::node::NodeExt;
use crate::lsp::traits::rope::RopeExt;
use crate::treesitter::NodeTypeExt;

pub fn goto_definition<'a>(
    document: &'a Document,
    params: GotoDefinitionParams,
) -> Result<Option<GotoDefinitionResponse>> {
    // get reference to AST
    let ast = &document.ast;

    let contents = &document.contents;

    // try to find node at position
    let position = params.text_document_position_params.position;
    let point = convert_position_to_point(contents, position);

    let Some(node) = ast.root_node().find_closest_node_to_point(point) else {
        log::warn!("Failed to find the closest node to point {point}.");
        return Ok(None);
    };

    let start = convert_point_to_position(contents, node.start_position());
    let end = convert_point_to_position(contents, node.end_position());
    let range = Range { start, end };

    // search for a reference in the document index
    if node.is_identifier() {
        let symbol = document.contents.node_slice(&node)?.to_string();
        if let Some((path, entry)) = indexer::find(symbol.as_str()) {
            let link = LocationLink {
                origin_selection_range: None,
                target_uri: Url::from_file_path(path).unwrap(),
                target_range: entry.range,
                target_selection_range: entry.range,
            };
            let response = GotoDefinitionResponse::Link(vec![link]);
            return Ok(Some(response));
        }
    }

    // TODO: We should see if we can find the referenced item in:
    //
    // 1. The document's current AST,
    // 2. The public functions from other documents in the project,
    // 3. A definition in the R session (which we could open in a virtual document)
    //
    // If we can't find a definition, then we can return the referenced item itself,
    // which will tell Positron to instead try to look for references for that symbol.
    let link = LocationLink {
        origin_selection_range: Some(range),
        target_uri: params.text_document_position_params.text_document.uri,
        target_range: range,
        target_selection_range: range,
    };

    let response = GotoDefinitionResponse::Link(vec![link]);
    Ok(Some(response))
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use assert_matches::assert_matches;
    use tower_lsp::lsp_types;

    use super::*;
    use crate::lsp::documents::Document;
    use crate::lsp::indexer;

    #[test]
    fn test_goto_definition() {
        let _guard = indexer::ResetIndexerGuard;

        let code = r#"
foo <- 42
print(foo)
"#;
        let doc = Document::new(code, None);
        let path = PathBuf::from("/foo/test.R");

        indexer::update(&doc, &path).unwrap();

        let params = GotoDefinitionParams {
            text_document_position_params: lsp_types::TextDocumentPositionParams {
                text_document: lsp_types::TextDocumentIdentifier {
                    uri: Url::from_file_path(&path).unwrap(),
                },
                position: lsp_types::Position::new(2, 7),
            },
            work_done_progress_params: Default::default(),
            partial_result_params: Default::default(),
        };

        assert_matches!(
            goto_definition(&doc, params).unwrap(),
            Some(GotoDefinitionResponse::Link(ref links)) => {
                assert!(!links.is_empty());
                assert_eq!(links[0].target_uri, Url::from_file_path(&path).unwrap());

                assert_eq!(
                    links[0].target_range,
                    lsp_types::Range {
                        start: lsp_types::Position::new(1, 0),
                        end: lsp_types::Position::new(1, 3),
                    }
                );
            }
        );
    }

    #[test]
    fn test_goto_definition_comment_section() {
        let _guard = indexer::ResetIndexerGuard;

        let code = r#"
# foo ----
foo <- 1
print(foo)
"#;
        let doc = Document::new(code, None);
        let path = PathBuf::from("/foo/section_test.R");
        indexer::update(&doc, &path).unwrap();

        let params = lsp_types::GotoDefinitionParams {
            text_document_position_params: lsp_types::TextDocumentPositionParams {
                text_document: lsp_types::TextDocumentIdentifier {
                    uri: lsp_types::Url::from_file_path(&path).unwrap(),
                },
                position: lsp_types::Position::new(3, 7),
            },
            work_done_progress_params: Default::default(),
            partial_result_params: Default::default(),
        };

        assert_matches!(
            goto_definition(&doc, params).unwrap(),
            Some(lsp_types::GotoDefinitionResponse::Link(ref links)) => {
                assert!(!links.is_empty());

                let link = &links[0];
                assert_eq!(
                    link.target_uri,
                    lsp_types::Url::from_file_path(&path).unwrap()
                );

                // The section should is not the target, the variable has priority
                assert_eq!(
                    links[0].target_range,
                    lsp_types::Range {
                        start: lsp_types::Position::new(2, 0),
                        end: lsp_types::Position::new(2, 3),
                    }
                );
            }
        );
    }
}
