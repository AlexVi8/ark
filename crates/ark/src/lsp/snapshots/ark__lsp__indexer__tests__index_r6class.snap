---
source: crates/ark/src/lsp/indexer.rs
expression: entries
---
[
    IndexEntry {
        key: "initialize",
        range: Range {
            start: Position {
                line: 3,
                character: 8,
            },
            end: Position {
                line: 3,
                character: 18,
            },
        },
        data: Method {
            name: "initialize",
        },
    },
    IndexEntry {
        key: "public_method",
        range: Range {
            start: Position {
                line: 6,
                character: 8,
            },
            end: Position {
                line: 6,
                character: 21,
            },
        },
        data: Method {
            name: "public_method",
        },
    },
    IndexEntry {
        key: "private_method",
        range: Range {
            start: Position {
                line: 11,
                character: 8,
            },
            end: Position {
                line: 11,
                character: 22,
            },
        },
        data: Method {
            name: "private_method",
        },
    },
    IndexEntry {
        key: "class",
        range: Range {
            start: Position {
                line: 1,
                character: 0,
            },
            end: Position {
                line: 1,
                character: 5,
            },
        },
        data: Variable {
            name: "class",
        },
    },
]
